Expose ansible_version to playbooks, add version_compare filter. Fixes #6583

sivel opened the pull request:
A number of people have asked for the ansible version to be exposed to playbooks.

This pull request exposes it as `ansible_version`.

Additionally, on it's own, it is complicated to compare that version to another value.

This pull request additionally adds a `version_compare` filter for easy comparison of versions.

Unit tests have been added for the new filter, and here is an example playbook to see this in action:

```

---
- hosts: all
  tasks:
    - debug:
        var: ansible_version

    - debug:
        msg: "{{ ansible_version|version_compare(1.6) }}"

    - debug:
        msg: "{{ ansible_version|version_compare(1.4, 'gt') }}"

    - debug:
        msg: "{{ ansible_version|version_compare(1.7, 'lt') }}"

    - debug:
        msg: "{{ ansible_version|version_compare(1.6, '>=') }}"

    - debug:
        msg: "{{ ansible_version|version_compare(1.6, '<') }}"
```

The `version_compare` accepts an 'operator' argument, which supports: <, lt, <=, le, >, gt, >=, ge, ==, =, eq, !=, <>, ne

sivel committed:
Expose ansible_version to playbooks, add version_compare filter to go along with it

mpdehaan commented:
This looks fine to me, though when adding a new filter plugin we also need to add it to the docs section about filters.

sivel commented:
I'll get the docs added shortly.

sivel commented:
I've added docs about `version_compare` and `ansible_version` and squashed the commits.

srgvg commented:
Appears I oversaw this PR when making my own :) See #6619 which however only introduces the ansible_version variable, and no Jinja filter, albeit:
- in runner, not in playbook, so it's also available outside of playbooks
- as a dict, where I split out the different components of the version string
- I used the code that generates the --version string used by the cli tools, which for a git checkout also includes extra information.

Maybe we could combine features?

sivel commented:
@sergevanginderachter  I'm not too convinced on the implementation in the other pull.  Also can you provide a use case for needing ansible_version outside of a playbook?

Also, I think that with a useful version_compare filter, a dictionary/hash value for ansible_version is potentially overcomplicated and unneeded.  If we think it is ultimately useful, a more simplistic approach would be something like:

```
version_string = utils.version('')
version = version_string.split()[0]
version_parts = map(int, ('%s.0' % version).split('.'))
inject['ansible_version'] = dict(release=version,
                                 major=version_parts[0],
                                 minor=version_parts[1],
                                 patch=version_parts[2],
                                 string=version_string)
```

@mpdehaan do you have any feedback?

srgvg commented:
@sivel Disregard my remark about availability outside playbook, I misread yours, you actually implemented at the same place as I did.

Your handing of integer conversion is obviously cleaner, but what if a version has a non-integer component? I was just making sure nothing would break for people that maintain a private branch, and might do versions like "2.3.0b1".

E.g. at one customer, I myself have this:

```
~# ansible --version
ansible 1.6.0-acd-20140312-0-custombugfixes
```

If your compare filter handles that too, fine by me.

mpdehaan commented:
I like the idea of using StrictVersion, while I don't agree it is quite necessary to have it available in Runner, that's consistent for most of Ansible provided it does not do calculuation each time.  Which I've commented in the other PR about.  So going to work on that one.

Sorry for duplicate efforts!

sivel commented:
Do you want a version_compare filter?  And are you saying that it should use StrictVersion instead?

I chose LooseVersion as that is what ansible-galaxy uses, and it is more flexible, but I can make that change.  I am more interested in the version_compare filter over exposing ansible_version anyway.

If you aren't interested, I won't waste time submitting another pull request.
